### PR TITLE
fix(oracle): set height to the next block of context when do recache

### PR DESCRIPTION
**Pull request by leonz789**
## Description

When a node restart, the oracle module will do refill the memory cache by replay some history blocks with executing the `PrepareNextRound`, `SealRound` process, the block height in context should be set to the corresponding history height.

## Changes
Notable changes:
- set block height of `context` to the corresponding history height when do recache
- refactor code when grow `roundID` with price unchanged
- add error log for case that `roundID` not match when append price for next round
----

Closes #XXX

## Summary by CodeRabbit

- **New Features**
  - Introduced fallback mechanisms for price aggregation failures, improving reliability in edge cases.
  - Enhanced round ID management in price aggregation to ensure continuity and accuracy.

- **Bug Fixes**
  - Improved error logging to assist in diagnosing issues more effectively.

- **Documentation**
  - Added comments in the code for better clarity and maintainability.

- **Refactor**
  - Renamed several functions for improved code readability and consistency.
  - Updated method calls in tests to align with new function names.

### CODE REVIEW

File: x/oracle/keeper/msg_server_create_price.go [hunk 1]
```diff
@@ -32,9 +32,14 @@ func (ms msgServer) CreatePrice(goCtx context.Context, msg *types.MsgCreatePrice
 		return &types.MsgCreatePriceResponse{}, nil
 	}
 	if newItem != nil {
-		ms.AppendPriceTR(ctx, newItem.TokenID, newItem.PriceTR)
-
-		logger.Info("final price aggregation done", "feederID", msg.FeederID, "roundID", newItem.PriceTR.RoundID, "price", newItem.PriceTR.Price)
+		if success := ms.AppendPriceTR(ctx, newItem.TokenID, newItem.PriceTR); !success {
+			// This case should not exist, keep this line to avoid consensus fail if this happens
+			// logger.Error("append new price round fail for mismatch roundID, and will just grow roundID with previous price", "roundID from finalPrice", newItem.PriceTR.RoundID)
```
Comment (cloud8little) [inline x/oracle/keeper/msg_server_create_price.go]: Ln37 is duplicated with Ln39.

Comment (leonz789) [reply x/oracle/keeper/msg_server_create_price.go]: Yes that was comment out. Removed.

**Comment by cloud8little** [comment]
Exocore version: PR #118 + PR #100 + PR #74 + PR #102 + PR #116

Setup 4 nodes as initial operators, initial voting power is: 2800, 2750, 2700, 1750 ; there are price feeders for node1 & node2
Stop node1 process at height **4699582** then start, shows consensus error message: 

node1 log:

```
8:11AM INF prepare for next oracle round of each tokenFeeder module=x/oracle
8:11AM INF executed block height=4699582 module=state num_invalid_txs=0 num_valid_txs=0 server=node
8:11AM INF commit synced commit=436F6D6D697449447B5B323133203633203130382031353320343420323220323437203433203739203130302032313320313137203136322031393320393120323230203136362032343820393720313030203132322034332031393220362036352031303220383720323820323120323120323338203139335D3A3437423542457D module=server
8:11AM INF committed state app_hash=D53F6C992C16F72B4F64D575A2C15BDCA6F861647A2BC0064166571C1515EEC1 height=4699582 module=state num_txs=0 server=node
8:11AM INF indexed block events height=4699582 module=txindex server=node
^C8:11AM INF HTTP server shut down, waiting 5 sec module=server
8:11AM INF RPC HTTP server stopped err="accept tcp 127.0.0.1:1317: use of closed network connection" module=server server=api
8:11AM INF service stop impl=Node module=server msg="Stopping Node service" server=node
8:11AM INF Stopping Node module=server server=node
8:11AM INF service stop impl=EventBus module=events msg="Stopping EventBus service" server=node
8:11AM INF service stop impl=PubSub module=pubsub msg="Stopping PubSub service" server=node
8:11AM INF service stop impl=IndexerService module=txindex msg="Stopping IndexerService service" server=node
8:11AM INF service stop impl="P2P Switch" module=p2p msg="Stopping P2P Switch service" server=node
8:11AM INF service stop impl=MConn{127.0.0.1:54276} module=p2p msg="Stopping MConnection service" peer=d7523b4c066719499ecb569d0fcff1122d6fc7c9@127.0.0.1:54276 server=node
8:11AM ERR Stopping peer for error err="read tcp 127.0.0.1:30000->127.0.0.1:54276: use of closed network connection" module=p2p peer="Peer{MConn{127.0.0.1:54276} d7523b4c066719499ecb569d0fcff1122d6fc7c9 in}" server=node
8:11AM INF service stop impl="Peer{MConn{127.0.0.1:54276} d7523b4c066719499ecb569d0fcff1122d6fc7c9 in}" module=p2p msg="Stopping Peer service" peer=d7523b4c066719499ecb569d0fcff1122d6fc7c9@127.0.0.1:54276 server=node
8:11AM ERR error while stopping peer error="already stopped" module=p2p server=node
8:11AM INF service stop impl=MConn{127.0.0.1:42506} module=p2p msg="Stopping MConnection service" peer=755b36cf9098759223fee20ee66673ee581ca37a@127.0.0.1:42506 server=node
8:11AM INF Reconnecting to peer addr=d7523b4c066719499ecb569d0fcff1122d6fc7c9@127.0.0.2:29999 module=p2p server=node
8:11AM ERR Stopping peer for error err="read tcp 127.0.0.1:30000->127.0.0.1:42506: use of closed network connection" module=p2p peer="Peer{MConn{127.0.0.1:42506} 755b36cf9098759223fee20ee66673ee581ca37a in}" server=node
8:11AM ERR error while stopping peer error="already stopped" module=p2p server=node
8:11AM INF service stop impl="Peer{MConn{127.0.0.1:42506} 755b36cf9098759223fee20ee66673ee581ca37a in}" module=p2p msg="Stopping Peer service" peer=755b36cf9098759223fee20ee66673ee581ca37a@127.0.0.1:42506 server=node
8:11AM INF Reconnecting to peer addr=755b36cf9098759223fee20ee66673ee581ca37a@127.0.0.3:29998 module=p2p server=node
8:11AM INF service stop impl=MConn{127.0.0.1:57284} module=p2p msg="Stopping MConnection service" peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453@127.0.0.1:57284 server=node
8:11AM ERR Stopping peer for error err="read tcp 127.0.0.1:30000->127.0.0.1:57284: use of closed network connection" module=p2p peer="Peer{MConn{127.0.0.1:57284} 0cba867433dd36bfaf500e45bdc6f5c9f3cb2453 in}" server=node
8:11AM INF service stop impl="Peer{MConn{127.0.0.1:57284} 0cba867433dd36bfaf500e45bdc6f5c9f3cb2453 in}" module=p2p msg="Stopping Peer service" peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453@127.0.0.1:57284 server=node
8:11AM ERR error while stopping peer error="already stopped" module=p2p server=node
8:11AM INF Reconnecting to peer addr=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453@127.0.0.4:29997 module=p2p server=node
8:11AM INF service stop impl=Reactor module=blockchain msg="Stopping Reactor service" server=node
8:11AM ERR Error stopping pool err="already stopped" module=blockchain server=node
8:11AM INF service stop impl=ConsensusReactor module=consensus msg="Stopping Consensus service" server=node
8:11AM INF service stop impl=ConsensusState module=consensus msg="Stopping State service" server=node
8:11AM INF service stop impl=TimeoutTicker module=consensus msg="Stopping TimeoutTicker service" server=node
8:11AM INF service stop impl=baseWAL module=consensus msg="Stopping baseWAL service" server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM ERR Won't start a peer - switch is not running module=p2p peer="Peer{MConn{127.0.0.1:54268} d7523b4c066719499ecb569d0fcff1122d6fc7c9 in}" server=node
8:11AM ERR Won't start a peer - switch is not running module=p2p peer="Peer{MConn{127.0.0.1:54282} 0cba867433dd36bfaf500e45bdc6f5c9f3cb2453 in}" server=node
8:11AM ERR Won't start a peer - switch is not running module=p2p peer="Peer{MConn{127.0.0.1:54288} 755b36cf9098759223fee20ee66673ee581ca37a in}" server=node
8:11AM INF service stop impl=Group module=consensus msg="Stopping Group service" server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM INF service stop impl=Evidence module=evidence msg="Stopping Evidence service" server=node
8:11AM INF service stop impl=StateSync module=statesync msg="Stopping StateSync service" server=node
8:11AM ERR Stopped accept routine, as transport is closed module=p2p numPeers=0 server=node
8:11AM INF Closing rpc listener listener={"Listener":{}} module=server server=node
8:11AM INF RPC HTTP server stopped err="accept tcp 127.0.0.1:20000: use of closed network connection" module=rpc-server server=node
8:11AM ERR Error serving server err="accept tcp 127.0.0.1:20000: use of closed network connection" module=server server=node
8:11AM INF Closing blockstore module=server server=node
8:11AM INF Closing statestore module=server server=node
8:11AM INF Closing evidencestore module=server server=node
[root@testnet-exochain tools]# exocored start --metrics  --log_level info --minimum-gas-prices=0.0001aexo --json-rpc.api eth,txpool,personal,net,debug,web3 --api.enable --json-rpc.enable true --grpc.enable true --home /root/.testnetV3/node1 --chain-id exocoretestnet_233-3
8:11AM INF Unlocking keyring module=server
8:11AM INF starting ABCI with Tendermint module=server
8:11AM INF starting node with ABCI Tendermint in-process module=server
8:11AM INF service start impl=multiAppConn module=proxy msg="Starting multiAppConn service" server=node
8:11AM INF service start connection=query impl=localClient module=abci-client msg="Starting localClient service" server=node
8:11AM INF service start connection=snapshot impl=localClient module=abci-client msg="Starting localClient service" server=node8:11AM INF service start connection=mempool impl=localClient module=abci-client msg="Starting localClient service" server=node
8:11AM INF service start connection=consensus impl=localClient module=abci-client msg="Starting localClient service" server=node
8:11AM INF service start impl=EventBus module=events msg="Starting EventBus service" server=node
8:11AM INF service start impl=PubSub module=pubsub msg="Starting PubSub service" server=node
8:11AM INF service start impl=IndexerService module=txindex msg="Starting IndexerService service" server=node
8:11AM INF ABCI Handshake App Info hash="�?l�,\x16�+Od�u��[ܦ�adz+�\x06AfW\x1c\x15\x15��" height=4699582 module=consensus protocol-version=0 server=node software-version=1.0.0-78-g7810930
8:11AM INF ABCI Replay Blocks appHeight=4699582 module=consensus server=node stateHeight=4699582 storeHeight=4699582
8:11AM INF Completed ABCI Handshake - CometBFT and App are synced appHash="�?l�,\x16�+Od�u��[ܦ�adz+�\x06AfW\x1c\x15\x15��" appHeight=4699582 module=consensus server=node
8:11AM INF Version info abci=1.0.0 block=11 commit_hash= module=server p2p=8 server=node tendermint_version=0.37.4
8:11AM INF This node is a validator addr=9D9F577A10F6C83B928C0268CA3C261CA4412FAE module=consensus pubKey=PubKeyEd25519{55AFFDD616D8C2F7FF15A8130A2EEC0445EFE958919BD9432D2735EE84ECF4EE} server=node
8:11AM INF P2P Node ID ID=3501b8bfd4aed1cfae1a828625933a6fa82955bb file=/root/.testnetV3/node1/config/node_key.json module=p2p
 server=node
8:11AM INF Adding persistent peers addrs=["d7523b4c066719499ecb569d0fcff1122d6fc7c9@127.0.0.2:29999","755b36cf9098759223fee20ee66673ee581ca37a@127.0.0.3:29998","0cba867433dd36bfaf500e45bdc6f5c9f3cb2453@127.0.0.4:29997"] module=p2p server=node8:11AM INF Adding unconditional peer ids ids=[] module=p2p server=node
8:11AM INF Add our address to book addr=3501b8bfd4aed1cfae1a828625933a6fa82955bb@127.0.0.1:30000 book=/root/.testnetV3/node1/config/addrbook.json module=p2p server=node
8:11AM INF Add our address to book addr=3501b8bfd4aed1cfae1a828625933a6fa82955bb@127.0.0.1:30000 book=/root/.testnetV3/node1/config/addrbook.json module=p2p server=node
8:11AM INF service start impl=Node module=server msg="Starting Node service" server=node
8:11AM INF Starting pprof server laddr=127.0.0.1:6060 module=server server=node
8:11AM INF service start impl="P2P Switch" module=p2p msg="Starting P2P Switch service" server=node
8:11AM INF service start impl=Reactor module=blockchain msg="Starting Reactor service" server=node
8:11AM INF service start impl=BlockPool module=blockchain msg="Starting BlockPool service" server=node
8:11AM INF serve module=rpc-server msg="Starting RPC HTTP server on 127.0.0.1:20000" server=node
8:11AM INF service start impl=ConsensusReactor module=consensus
 msg="Starting Consensus service" server=node
8:11AM INF Reactor  module=consensus server=node waitSync=true
8:11AM INF service start impl=Evidence module=evidence msg="Starting Evidence service" server=node
8:11AM INF service start impl=StateSync module=statesync msg="Starting StateSync service" server=node
8:11AM INF Saving AddrBook to file book=/root/.testnetV3/node1/config/addrbook.json module=p2p server=node size=3
8:11AM INF starting API server... module=server server=api
8:11AM INF serve module=server msg="Starting RPC HTTP server on 127.0.0.1:1317" server=api
8:11AM INF service start impl="Peer{MConn{127.0.0.2:29999} d7523b4c066719499ecb569d0fcff1122d6fc7c9 out}" module=p2p msg="Starting Peer service" peer=d7523b4c066719499ecb569d0fcff1122d6fc7c9@127.0.0.2:29999 server=node
8:11AM INF service start impl=MConn{127.0.0.2:29999} module=p2p msg="Starting MConnection service" peer=d7523b4c066719499ecb569d0fcff1122d6fc7c9@127.0.0.2:29999 server=node
8:11AM INF service start impl="Peer{MConn{127.0.0.3:29998} 755b36cf9098759223fee20ee66673ee581ca37a out}" module=p2p msg="Starting Peer service" peer=755b36cf9098759223fee20ee66673ee581ca37a@127.0.0.3:29998 server=node
8:11AM INF service start impl=MConn{127.0.0.3:29998} module=p2p
 msg="Starting MConnection service" peer=755b36cf9098759223fee20ee66673ee581ca37a@127.0.0.3:29998 server=node
8:11AM INF Time to switch to consensus reactor! height=4699583 module=blockchain server=node
8:11AM INF service stop impl=BlockPool module=blockchain msg="Stopping BlockPool service" server=node
8:11AM INF SwitchToConsensus module=consensus server=node
8:11AM INF service start impl=ConsensusState module=consensus msg="Starting State service" server=node
8:11AM INF service start impl=baseWAL module=consensus msg="Starting baseWAL service" server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM INF service start impl=Group module=consensus msg="Starting Group service" server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM INF service start impl=TimeoutTicker module=consensus msg="Starting TimeoutTicker service" server=node
8:11AM INF Searching for height height=4699583 max=0 min=0 module=consensus server=node wal=/root/.testnetV3/node1/data/cs.wal/wal8:11AM INF Searching for height height=4699582 max=0 min=0 module=consensus server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM INF Found height=4699582 index=0 module=consensus server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM INF Catchup by replaying consensus messages height=4699583 module=consensus server=node
8:11AM INF Replay: New Step height=4699583 module=consensus round=0 server=node step=RoundStepNewHeight
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582 module=consensus peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453
 round=0 server=node type=SIGNED_MSG_TYPE_PREVOTE
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582
 module=consensus peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582
 module=consensus peer=d7523b4c066719499ecb569d0fcff1122d6fc7c9
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582 module=consensus peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453
 round=0 server=node type=SIGNED_MSG_TYPE_PREVOTE
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582
 module=consensus peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582
 module=consensus peer=d7523b4c066719499ecb569d0fcff1122d6fc7c9
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582 module=consensus peer=d7523b4c066719499ecb569d0fcff1122d6fc7c9
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582 module=consensus peer=755b36cf9098759223fee20ee66673ee581ca37a
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582
 module=consensus peer=755b36cf9098759223fee20ee66673ee581ca37a
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582
 module=consensus peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Vote blockID=105E8B88539698BE46BCD45452C085548558D070CF2CC8C0F7822479E0F3F78D:1:3E3A9A3C2B4F height=4699582 module=consensus peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453
 round=0 server=node type=SIGNED_MSG_TYPE_PRECOMMIT
8:11AM INF Replay: Done module=consensus server=node8:11AM INF Replay: Done module=consensus server=node
8:11AM INF commit is for a block we do not know about; set ProposalBlock=nil commit=D665C097A1510DF1F9F414851FACA7C769223AF469B95BD05AC8FF66F870B4C6 commit_round=0 height=4699583 module=consensus proposal= server=node
8:11AM INF received complete proposal block hash=D665C097A1510DF1F9F414851FACA7C769223AF469B95BD05AC8FF66F870B4C6 height=4699583 module=consensus server=node
8:11AM INF finalizing commit of block hash=D665C097A1510DF1F9F414851FACA7C769223AF469B95BD05AC8FF66F870B4C6 height=4699583 module=consensus num_txs=1 root=D53F6C992C16F72B4F64D575A2C15BDCA6F861647A2BC0064166571C1515EEC1 server=node
8:11AM INF add price proposal for aggregation basedBlock=4699580 feederID=1 module=x/oracle proposer=exo1nk04w7ss7myrhy5vqf5v50pxrjjyztawfy3tj0
8:11AM INF add new round with previous price under fail aggregation module=x/oracle price= roundID=39 tokenID=1
8:11AM INF prepare for next oracle round of each tokenFeeder module=x/oracle
8:11AM INF executed block height=4699583 module=state num_invalid_txs=0 num_valid_txs=1 server=node
8:11AM INF commit synced commit=436F6D6D697449447B5B3338203136352032353120383220363120323333203132302031393620323039203135342034362031353320313133203633203139382034372039332035203134302032333820373920333520373320393820343020313430203739203436203532203234392037362033305D3A3437423542467D module=server8:11AM INF committed state app_hash=26A5FB523DE978C4D19A2E99713FC62F5D058CEE4F234962288C4F2E34F94C1E height=4699583 module=state num_txs=1 server=node
8:11AM INF indexed block events height=4699583 module=txindex server=node
8:11AM INF service start impl="Peer{MConn{127.0.0.4:29997} 0cba867433dd36bfaf500e45bdc6f5c9f3cb2453 out}" module=p2p msg="Starting Peer service" peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453@127.0.0.4:29997 server=node
8:11AM INF service start impl=MConn{127.0.0.4:29997} module=p2p msg="Starting MConnection service" peer=0cba867433dd36bfaf500e45bdc6f5c9f3cb2453@127.0.0.4:29997 server=node
8:11AM INF received proposal module=consensus proposal="Proposal{4699584/0 (A4A0459545676636C8873F17814C280FCC1B4E70542137160410A2B2938EC02B:1:F8F0E69C49F5, -1) 79CE57A75BEA @ 2024-07-02T08:11:33.708443244Z}" proposer=2F703B396E582FB223561E445610EEFCFC0E594C server=node
8:11AM INF received complete proposal block hash=A4A0459545676636C8873F17814C280FCC1B4E70542137160410A2B2938EC02B height=4699584 module=consensus server=node8:11AM ERR prevote step: consensus deems this block invalid; prevoting nil err="wrong Block.Header.LastResultsHash.  Expected 877D02DFCF82AD7F31ABA3017468EFDF642079EFA7CDBA9BB6C12F9333FD3D29, got 9B0844BB418DB18FB4DC7F6B7A36037FF7455AA5BE530D6386C86E123519C47A" height=4699584 module=consensus round=0 server=node
8:11AM ERR CONSENSUS FAILURE!!! err="precommit step; +2/3 prevoted for an invalid block: wrong Block.Header.LastResultsHash.  Expected 877D02DFCF82AD7F31ABA3017468EFDF642079EFA7CDBA9BB6C12F9333FD3D29, got 9B0844BB418DB18FB4DC7F6B7A36037FF7455AA5BE530D6386C86E123519C47A" module=consensus server=node stack="goroutine 128 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x5e\ngithub.com/cometbft/cometbft/consensus.(*State).receiveRoutine.func2()\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:736 +0x46\npanic({0x2cb0920?, 0xc0013fd260?})\n\truntime/panic.go:770 +0x132\ngithub.com/cometbft/cometbft/consensus.(*State).enterPrecommit(0xc001659c08, 0x47b5c0, 0x0)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1463 +0x1599\ngithub.com/cometbft/cometbft/consensus.(*State).addVote(0xc001659c08, 0xc001c50e60, {0xc006d66360, 0x28})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:2195 +0x138f\ngithub.com/cometbft/cometbft/consensus.(*State).tryAddVote(0xc001659c08, 0xc001c50e60, {0xc006d66360?, 0x0?})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:2012 +0x26\ngithub.com/cometbft/cometbft/consensus.(*State).handleMsg(0xc001659c08, {{0x42b1520, 0xc001c342b0}, {0xc006d66360, 0x28}})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:865 +0x39c\ngithub.com/cometbft/cometbft/consensus.(*State).receiveRoutine(0xc001659c08, 0x0)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:772 +0x3f1\ncreated by github.com/cometbft/cometbft/consensus.(*State).OnStart in goroutine 115\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:383 +0x10c\n"
8:11AM INF service stop impl=baseWAL module=consensus msg="Stopping baseWAL service" server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM INF service stop impl=Group module=consensus msg="Stopping Group service" server=node wal=/root/.testnetV3/node1/data/cs.wal/wal
8:11AM INF Timed out dur=1000 height=4699584 module=consensus round=0 server=node step=RoundStepPrevoteWait
8:11AM INF Transactions per second module=server tps=0.1
8:11AM INF New websocket connection module=rpc-server protocol=websocket remote=127.0.0.1:48464 server=node
```

node1 price feeder log:

```
2024/07/02 08:11:23 update price:343361940000, decimal:8
2024/07/02 08:11:27 debug-newblock-height: 4699582
2024/07/02 08:11:27 debug-feeder. triggered, feeder-parames:{feederID:1, startBlock:30, interval:10, roundID:1}
2024/07/02 08:11:27 submit price=343361940000 of token=ETHUSDT on height=4699582 for roundID:469956
exo1nk04w7ss7myrhy5vqf5v50pxrjjyztawfy3tj0
2024/07/02 08:11:27 sendTx successed
read err
panic: websocket: close 1006 (abnormal closure): unexpected EOF
goroutine 28 [running]:
github.com/ExocoreNetwork/price-feeder/exoclient.Subscriber.func2()
        /root/Jessica/Github/price-feeder/exoclient/subscribe.go:79 +0x278
created by github.com/ExocoreNetwork/price-feeder/exoclient.Subscriber in goroutine 1
        /root/Jessica/Github/price-feeder/exoclient/subscribe.go:66 +0x2a6
exit status 2
```

**Comment by cloud8little** [comment]
The key reproduction step is shutting down the node service right after the price feeder submit a create-price message transaction, and then start, the `CONSENSUS FAILURE` error message will show.

**Comment by leonz789** [comment]
fixed by #122

**Comment by cloud8little** [comment]
Test pass with https://github.com/ExocoreNetwork/exocore/pull/118/commits/67eaf49659f352c454bc0fc46362e1689de1fb42.

Restart price feeder and make a create-price transaction then shut off the node, start the node again, it works well to sync and generate new blocks.

**Review by cloud8little — APPROVED** [review]

**Review by bwhour — APPROVED** [review]

**Review by mikebraver — APPROVED** [review]
LGTM

**Comment by mikebraver** [comment]
@leonz789  Can be merged after resolving the action issue

**Pull request merged**